### PR TITLE
Nuke op voucher system and uplink changes

### PR DESCRIPTION
**Pull request by ghost**
- Nuke ops no longer have bulldog shotguns by default in their team locker.
- Pinpointers moved to personal lockers.
- Syndicate pistol now comes with suppressor by default, is traitor/DA only (nuke ops need to buy the suppressor)
- Nuke ops can no longer buy revolvers
- Nuke ops can no longer buy C-20rs
- Nuke ops start with a weapon voucher by default to choose their starting weapon. Starting weapons come with an extra magazine.
- Hit uplink with voucher to redeem.
- Nuke ops can redeem voucher for an extra 3 telecrystals should they not want a starting weapon.
- Slightly nerfed c20r stamina damage

**List of starter weapons from the voucher:**
- C-20r SMG
- Bulldog shotgun
- Single-shot break-action grenade launcher (ala m79)
- Surplus mosin-nagant

jizz
someone can probably make this better than me

### CODE REVIEW

**Comment by WJohn** [comment]
Interesting idea, should help prevent people from buying a dozen guns they can't ever use, at least. They can still buy the ammo for these weapons in the uplink, right?

**Comment by paprka** [comment]
Yeah

**Comment by Razharas** [comment]
Now THIS definitely needs a detailed changelog, or nuke ops rounds a few weeks from now will be full of rage, tears and regrets

**Comment by paprka** [comment]
Good, i exist to torment antags

**Comment by Fox-McCloud** [comment]
Not much of a reason to ever go with bulldogs in this case. Dragon's breath rounds are useless, and with the burst fire system of the C20r, it's going to be preferable to the bulldog stunshells in most case...leaving only buckshot for multiple grouped targets...only with how erratically it behaves and the fact most combat doesn't typically benefit buckshot---well....I honestly don't expect to see the bulldog much anymore if this gets merged in.

**Comment by paprka** [comment]
I don't really plan on doing anything special here in regards to gun balance, if people take the easy route with the c20r I don't really have a problem with that, it's a classic nuke op gun with a lot of power in a compact and balanced package

**Comment by Steelpoint** [comment]
Interesting, but as noted people will only get a C20r as its better than the Bulldog in all respects. Imo the Bulldog was a "budget" C20 as you could spend a few TC's to give it a few stun/buckshot rounds as a makeshift C20.

Also will you still need to buy the C90?

**Comment by paprka** [comment]
Yeah, and I don't really have any problems with that, it's essentially a c-20r + grenade launcher for 18tc, which is a lot

**Comment by Dorsisdwarf** [comment]
So it's essentially a grenade launcher for 18TC then? Since the CR-20 can now be your free starter gun.

**Comment by paprka** [comment]
Eh? no, it has a massive magazine in comparison, and does 10 more damage with each shot, as well as having a longer burst fire, in addition to having a grenade launcher attached for easy usage. It's essentially the middle ground between the c-20r and the l6, still affordable for normal ops but it eats up a lot of TC.

**Comment by Dorsisdwarf** [comment]
I've never seen the C-90 , I was going off your statement that it's a c20r + grenade launcher.

**Comment by paprka** [comment]
I misspoke, it's much better, and I dont' want to remove it (even though I should) because it was a pretty successful venture into guncode and contained attachments

**Comment by AlexanderUlanH** [comment]
If we're going to start putting more vouchers in the game, the redemption code should be moved to the voucher so it can be done properly.  I know the response is going to be 
>If you want my code take it as-is, I can't be bothered to code my features properly.

But every time this stuff gets merged, our code gets less maintainable.

**Comment by paprka** [comment]
Why would that be 'done properly'

Why would this make the code less maintainable?

**Comment by AlexanderUlanH** [comment]
Done properly, as opposed to code being copied and pasted from mining vouchers and equipment lockers to syndie vouchers and uplinks.

**Comment by paprka** [comment]
What is done properly? Can you give me an example or are you just hating copy paste for entirely the wrong reasons even though the code works perfectly fine?

Oh, my mistake, you must like to start shit :banana:

**Comment by AlexanderUlanH** [comment]
Ninja code also works.  I'll give you the benefit of the doubt here:
http://lmgtfy.com/?q=why+copypaste+code+is+bad
I said for fairness' sake that "done properly" means moving the redemption code to the voucher, but I'm appealing to the maintainers, because I don't expect you to take an interest in improving your code, and everyone's free to (and typically does) contribute in ways that interest them.

**Comment by paprka** [comment]
The redemption code shouldn't be on the voucher, it's only for the nuke op uplink which is already a subtype and is already pretty much used exclusively for the icon alone. The item itself is worthless outside of being used on the uplink, so why move the procs to the uplink? Just because it's too similar to the mining voucher's system?

Lol.

You can at least pretend to know what the point of code is that you're criticizing, but feel free to contribute to making yourself look like a moron.

**Comment by AlexanderUlanH** [comment]
I'll readily admit that I'm a moron, and if there's a better way to make redemption vouchers polymorphic, I'm sorry for offering a stupid suggestion.  My point is that the code as it stands shouldn't be merged because it uses copypasted code as a substitute for polymorphism, which is bad enough that it was included in the standards.
https://tgstation13.org/wiki/Coding_Standards

**Comment by paprka** [comment]
:laughing: Holy shit you are fucking stupid. It is an extremely simple proc with a switch. Seriously, did you even look at the code? Next you'll tell me all switches are copypaste.

**Comment by AlexanderUlanH** [comment]
If the maintainers want to make an exception because it's not much code, I disagree, but that's on them.  There's more to the code than a switch statement, and it was copied from code/modules/mining/equipment_locker.dm, and there's an easy polymorphic way to avoid copypasting the code, and copypasting code when there's an easy polymorphic way to avoid doing so is against standards, so this shouldn't be merged.

**Comment by paprka** [comment]
"and there's an easy polymorphic way to avoid copypasting the code"

I'm waiting.

**Comment by AlexanderUlanH** [comment]
I told you already here, and once on coderbus for good measure.

**Comment by paprka** [comment]
Yes but you didn't suggest how to go about it

Both vouchers would need overrides which would take up the same amount of code in the end, for a system that really doesn't have the potential of modularity to support it.

You're complaining just to complain, in other words.

File: code/game/objects/items/devices/uplinks.dm [hunk 1]
```diff
@@ -211,5 +211,35 @@ var/list/world_uplinks = list()
 	hidden_uplink = new(src)
 	hidden_uplink.uses = 20
 
+/obj/item/device/radio/uplink/attackby(obj/item/I as obj, mob/user as mob)
+	if(istype(I, /obj/item/weapon/syndie_voucher))
+		RedeemVoucher(I, user)
+		return
 
-
+/obj/item/device/radio/uplink/proc/RedeemVoucher(obj/item/weapon/syndie_voucher/voucher, mob/redeemer)
```
Comment (Cheridan) [inline code/game/objects/items/devices/uplinks.dm]: He's right. There are more kinds of uplinks than radios. The voucher should simply check if the item it's hitting has an uplink (check TC console code for an example) and work from there.

Comment (paprka) [inline code/game/objects/items/devices/uplinks.dm]: This isn't a good way to go about it since it only applies to the nuke op uplink intentionally

**Comment by Steelpoint** [comment]
I think this is a cool system, but I think there are going to be several flaws with the system, I would argue we should rename this from "Nuke op voucher system" to "All Ops spawn with a C20r".

All of available weapons on offer are, subjectively, inferior to the C20r. The Mosin-Naget is a gimmick, the single shot nade launcher is interesting but not powerful enough and the Bulldog is, again, not powerful enough to match the power of the C20r. There is no reason to NOT take the C20r for free.

What's of additional concern are two possible knock on effects, them being...

* Proliferation of Energy Shields
* No one buying the C90 Assault Rifle

Energy Shields are insanely robust, the only reason they are not commonly used is because they are very expensive to buy for each operative, however now that suddenly all the Ops have a free gun they can all now afford to buy a EShield.

Secondly, the C90, while a good gun, is not good enough to justify expending 18 TC's on it when you can get a free C20 and have a extra 18 TC's to spend on something else (like a EShield)

-----

I think what should be done is that the C90 Assault Rifle is made available as one of the possible starting weapons, but we remove its grenade launcher attachment and make that something you have to buy.

Also look into giving the Bulldog more robust ammo.

**Comment by Fox-McCloud** [comment]
@Steelpoint 

I'm pretty sure Steelpoint is completely on the money on this one. From my experience, c20Rs are, by far, the most commonly used nuke ops weapon at the moment, with backup buckshot being the fallback; Steel's point on energy shields is well grounded--an Op with an energy shield is only impacted by bullets or polyacid-smoke darts.

From my experience, usually there's only one (and sometimse two) ops who goes with a shield and typically has a stun weapon to be the teams "tank" for the front lines. If every Op suddenly gets a shield, there just won't be much anyone can do to stop the ops; two riot shotguns, a double-barrel, and the detective's revolver isn't enough to control a team of flashbang and energy+laser immune targets--even more so once Ops figure out that breaching at the armory not only allows them to have access to more weaponry, but eleminates 2 of the only 4 threats to them on the station.

I like the voucher system, but the weapons available should be on a similar power level---when one of the mainstays of nukies is suddenly available for free to the team, it's essentially just saying "here, have a free 14 TC for everyone".

**Comment by Jordie0608** [comment]
Changelog

**Pull request closed without merging**